### ISSUE TITLE
LGBMRanker does not call custom objective function with group

### ISSUE THREAD

**Issue by francoisWeber** [microsoft/LightGBM#7018]
## Description
I tried a custom ranking objective function for LGBMRanker but LGBMRanker does not call the custom loss function with the `group` argument, which is mandatory for learning to rank.

## Reproducible example
```
import lightgbm as lgbm
import numpy as np

# fake data
N, p, gsize = 1000, 10, 5
X = np.random.randn(N, p)
y = np.random.randint(0, 4, N)
groups = [gsize] * (N//gsize)


# fake loss to trace inputs 
_args = None 
_kwargs = None
def loss(*args, **kwargs):
    global _args, _kwargs
    _args = args
    _kwargs = kwargs
    return 0, 0


lgbm_ranker = lgbm.LGBMRanker(n_estimators=10, learning_rate=0.1, objective=loss)
try:
    lgbm_ranker.fit(X, y, group=groups)
except Exception as e:
    print(e)
print(len(_args))
print(_kwargs)
```

`_args` nor `_kwargs` contain reference to the `groups` I used as ranking group size.

## Environment info

LightGBM version or commit hash: 4.6.0

Command(s) you used to install LightGBM

```shell
poetry add lightgbm
```

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

For the `scikit-learn` APIs, LightGBM determines how many arguments to pass to custom objective functions by inspecting the signatures of the functions.

https://github.com/microsoft/LightGBM/blob/de4c88f92c619fed7d4d61056b6c2b8b41107691/python-package/lightgbm/sklearn.py#L221

https://github.com/microsoft/LightGBM/blob/de4c88f92c619fed7d4d61056b6c2b8b41107691/python-package/lightgbm/sklearn.py#L231-L233

Maybe something else should be done there to handle `*args, **kwargs` in signatures, but that'd require some design work.

For now, please follow the exact signature indicated in the docs:

https://github.com/microsoft/LightGBM/blob/de4c88f92c619fed7d4d61056b6c2b8b41107691/python-package/lightgbm/sklearn.py#L162-L188

And also note that the docs there say the gradient and hessians should be 1D arrays (not a tuple of ints as your example is using).

**Comment by francoisWeber** [comment]
OK thanks a lot, @jameslamb - the code is running now. However I must have done mistakes in it because the training's behavior is wrong. 

I give up for now but let me share my attempts with others. Here is what I tried:
- use [RAX](https://rax.readthedocs.io/)'s smoothed ranking loss to have more LTR losses available in LGBM (e.g: optimizing on recall)
- use JAX (which is used by RAX) to obtain gradients and hessians 
- plug them into a custom loss 

My WIP is:
```python
# Reimplement NDCG loss for LightGBM through RAX API

import rax
from functools import partial
import jax
import numpy as np
from typing import Tuple
from enum import Enum

class RankingLoss(Enum):
    """Loss functions for ranking"""
    NDCG = rax.ndcg_metric
    MRR = rax.mrr_metric
    RECALL = rax.recall_metric
    PRECISION = rax.precision_metric
    

class RankingLossForLGBM:
    def __init__(self, loss: RankingLoss, topn: int = 10):
        self.topn = topn
        self._loss = partial(rax.approx_t12n(loss), topn=topn)
        self._grad = jax.grad(self._loss)
        self._hess = jax.hessian(self._loss)
        
    def _call_on_one_group(self, labels: np.ndarray, preds: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
        return self._grad(preds, labels), np.diag(self._hess(preds, labels))

    def __call__(self, labels: np.ndarray, preds: np.ndarray, weights: np.ndarray = None, groups: list[int] | None = None) -> Tuple[np.ndarray, np.ndarray]:
        # Split labels into chunks of size grp for grp in groups
        groups = groups if groups is not None else [len(labels)]
        weights = weights or np.ones(len(groups))
        start = 0
        grads = []
        hessians = []
        for grp, w in zip(groups, weights):
            end = start + grp
            _labels = labels[start:end]
            _preds = preds[start:end]
            grad, hess = self._call_on_one_group(_preds, _labels)
            grads.append(grad * w)
            hessians.append(hess * w)
            start = end
        return np.hstack(grads), np.hstack(hessians)


# fake data
N, p, gsize = 1000, 10, 5
X = np.random.randn(N, p)
y = np.random.randint(0, 4, N)
groups = [gsize] * (N//gsize)

# training
loss = RankingLossForLGBM(RankingLoss.NDCG, topn=2)
lgbm_ranker = lgbm.LGBMRanker(n_estimators=10, learning_rate=0.1, objective=loss)
lgbm_ranker.fit(X, y, group=groups)
```

I give up for now but this could give ideas to others.
